### PR TITLE
support `rvar` columns in data.frames

### LINKED ISSUE #604: Add support for data frames with an `rvar` column
The `rvar` class is a super convenient way to wrangle posteriors. 

Here is a basic workflow I've adopted:

```r
library(brms)
library(posterior)
library(bayestestR)

mod <- brm(mpg ~ factor(cyl) + hp, data = mtcars,
           backend = "cmdstanr", cores = 4, refresh = 0)
#> Compiling Stan program...
#> Start sampling
#> Running MCMC with 4 parallel chains...
#> 
#> Chain 1 finished in 0.3 seconds.
#> Chain 2 finished in 0.3 seconds.
#> Chain 3 finished in 0.2 seconds.
#> Chain 4 finished in 0.2 seconds.
#> 
#> All 4 chains finished successfully.
#> Mean chain execution time: 0.3 seconds.
#> Total execution time: 0.5 seconds.
#> 

grid <- expand.grid(
  cyl = unique(mtcars$cyl),
  hp = mean(mtcars$hp)
)

grid$epred <- posterior_epred(mod, grid) |> rvar()

grid
#>   cyl       hp    epred
#> 1   6 146.6875 19 ± 1.3
#> 2   4 146.6875 25 ± 1.4
#> 3   8 146.6875 17 ± 1.3
```

Unfortunately, this data frame structure does not work with `bayestestR` functions (here is an example with `describe_posterior()`):

```r
describe_posterior(grid)
#> Error in rbind(deparse.level, ...) : 
#>   numbers of columns of arguments do not match
```

And passing the `rvar` column loses all the grid data:

```r
describe_posterior(grid$epred)
#> Summary of Posterior Distribution
#> 
#> Parameter | Median |         95% CI |   pd |          ROPE | % in ROPE
#> ----------------------------------------------------------------------
#> x[1]      |  19.13 | [16.63, 21.74] | 100% | [-0.10, 0.10] |        0%
#> x[2]      |  25.11 | [22.24, 27.94] | 100% | [-0.10, 0.10] |        0%
#> x[3]      |  16.57 | [14.00, 19.26] | 100% | [-0.10, 0.10] |        0%
```

What I'd like is for the `foo.data.frame()` methods to check if there is an `rvar` column - if there is (only 1), it should be processed and added to the grid. Like so:

```r
describe_posterior.data.frame <- function(posteriors, ...) {
  if (any(is_rvar <- sapply(posteriors, inherits, "rvar"))) {
    if (sum(is_rvar, na.rm = TRUE) > 1L) stop("Cannot use more than 1 rvar")
    
    cl <- match.call()
    cl[[1]] <- quote(describe_posterior)
    cl$posteriors <- posteriors[,is_rvar, drop = TRUE]
    out <- eval(cl)
    out$Parameter <- NULL
    
    other_columns <- posteriors[,!is_rvar, drop = FALSE]
    to_add <- colnames(other_columns)
    out[to_add] <- other_columns
    out <- datawizard::data_relocate(out, select = to_add, before = 1)
    return(out)
  }

  # else, do as usual...
  bayestestR:::describe_posterior.data.frame(posteriors, ...)
}

describe_posterior(grid)
#> Summary of Posterior Distribution
#> 
#> cyl  |     hp | Median |         95% CI |   pd |          ROPE | % in ROPE
#> --------------------------------------------------------------------------
#> 6.00 | 146.69 |  19.19 | [16.50, 21.88] | 100% | [-0.10, 0.10] |        0%
#> 4.00 | 146.69 |  25.16 | [22.28, 27.99] | 100% | [-0.10, 0.10] |        0%
#> 8.00 | 146.69 |  16.55 | [13.94, 19.21] | 100% | [-0.10, 0.10] |        0%
```

Should be easy enough to implement...

### PR DESCRIPTION
**Pull request by mattansb**
Closes #604 

``` r
``` r
library(bayestestR)

dfx <- data.frame(mu = c(0, 0.5, 1), sigma = c(1, 0.5, 0.25))
dfx$my_rvar <- posterior::rvar_rng(rnorm, 3, mean = dfx$mu, sd = dfx$sigma)
dfx$other_rvar <- posterior::rvar_rng(rnorm, 3, mean = dfx$mu + 0.5, sd = dfx$sigma - 0.1)
dfx
#>    mu sigma     my_rvar  other_rvar
#> 1 0.0  1.00 0.01 ± 0.99 0.49 ± 0.91
#> 2 0.5  0.50 0.51 ± 0.49 1.00 ± 0.40
#> 3 1.0  0.25 1.00 ± 0.25 1.50 ± 0.15
```

``` r

map_estimate(dfx, rvar_col = "my_rvar")
#> MAP Estimate
#> 
#> mu   | sigma | MAP_Estimate
#> ---------------------------
#> 0.00 |  1.00 |         0.12
#> 0.50 |  0.50 |         0.52
#> 1.00 |  0.25 |         1.00
```

``` r

point_estimate(dfx, rvar_col = "my_rvar")
#> Point Estimate
#> 
#> mu   | sigma |   Median | Mean |  MAP
#> -------------------------------------
#> 0.00 |  1.00 | 6.21e-03 | 0.01 | 0.12
#> 0.50 |  0.50 |     0.50 | 0.51 | 0.52
#> 1.00 |  0.25 |     1.00 | 1.00 | 1.00
```

``` r

eti(dfx, rvar_col = "my_rvar")
#> Equal-Tailed Interval
#> 
#> mu   | sigma |       95% ETI
#> ----------------------------
#> 0.00 |  1.00 | [-1.87, 1.99]
#> 0.50 |  0.50 | [-0.44, 1.46]
#> 1.00 |  0.25 | [ 0.52, 1.48]
```

``` r

bci(dfx, rvar_col = "my_rvar")
#>    mu sigma   CI     CI_low  CI_high
#> 1 0.0  1.00 0.95 -1.8677618 1.999030
#> 2 0.5  0.50 0.95 -0.4308229 1.464526
#> 3 1.0  0.25 0.95  0.5241453 1.489345
```

``` r

hdi(dfx, rvar_col = "my_rvar")
#> Highest Density Interval
#> 
#> mu   | sigma |       95% HDI
#> ----------------------------
#> 0.00 |  1.00 | [-1.81, 2.03]
#> 0.50 |  0.50 | [-0.46, 1.43]
#> 1.00 |  0.25 | [ 0.52, 1.47]
```

``` r

spi(dfx, rvar_col = "my_rvar")
#> Shortest Probability Interval
#> 
#> mu   | sigma |       95% SPI
#> ----------------------------
#> 0.00 |  1.00 | [-1.78, 2.07]
#> 0.50 |  0.50 | [-0.46, 1.43]
#> 1.00 |  0.25 | [ 0.51, 1.47]
```

``` r

ci(dfx, rvar_col = "my_rvar")
#> Equal-Tailed Interval
#> 
#> mu   | sigma |       95% ETI
#> ----------------------------
#> 0.00 |  1.00 | [-1.87, 1.99]
#> 0.50 |  0.50 | [-0.44, 1.46]
#> 1.00 |  0.25 | [ 0.52, 1.48]
```

``` r

p_direction(dfx, rvar_col = "my_rvar")
#> Probability of Direction
#> 
#> mu   | sigma |     pd
#> ---------------------
#> 0.00 |  1.00 | 50.30%
#> 0.50 |  0.50 | 84.78%
#> 1.00 |  0.25 |   100%
```

``` r

p_map(dfx, rvar_col = "my_rvar")
#> MAP-based p-value
#> 
#> mu   | sigma | p (MAP)
#> ----------------------
#> 0.00 |  1.00 |  0.997 
#> 0.50 |  0.50 |  0.634 
#> 1.00 |  0.25 |  < .001
```

``` r

p_rope(dfx, rvar_col = "my_rvar", range = c(-1, 1))
#> Proportion of samples inside the ROPE [-1.00, 1.00]
#> 
#> mu   | sigma | p (ROPE)
#> -----------------------
#> 0.00 |  1.00 |    0.685
#> 0.50 |  0.50 |    0.837
#> 1.00 |  0.25 |    0.501
```

``` r

p_significance(dfx, rvar_col = "my_rvar", threshold = 0.5)
#> Practical Significance (threshold: 0.50)
#> 
#> mu   | sigma |   ps
#> -------------------
#> 0.00 |  1.00 | 0.31
#> 0.50 |  0.50 | 0.50
#> 1.00 |  0.25 | 0.98
```

``` r

rope(dfx, rvar_col = "my_rvar", range = c(-1, 0.5))
#> # Proportion of samples inside the ROPE [-1.00, 0.50]:
#> 
#>   mu | sigma | inside ROPE
#> --------------------------
#> 0.00 |  1.00 |     55.89 %
#> 0.50 |  0.50 |     49.79 %
#> 1.00 |  0.25 |      0.00 %
```

``` r

equivalence_test(dfx, rvar_col = "my_rvar", range = c(-1, 0.5), ci = 0.8)
#> # Test for Practical Equivalence
#> 
#>   ROPE: [-1.00 0.50]
#> 
#>   mu | sigma |        H0 | inside ROPE |      80% HDI
#> -----------------------------------------------------
#> 0.00 |  1.00 | Undecided |     66.38 % | [-1.26 1.32]
#> 0.50 |  0.50 | Undecided |     49.75 % | [-0.13 1.14]
#> 1.00 |  0.25 |  Rejected |      0.00 % | [ 0.68 1.32]
```

``` r

estimate_density(dfx, rvar_col = "my_rvar") |> head()
#>   mu sigma         x            y
#> 1  0     1 -3.566836 0.0007033624
#> 2  0     1 -3.560231 0.0007158896
#> 3  0     1 -3.553626 0.0007284776
#> 4  0     1 -3.547021 0.0007411670
#> 5  0     1 -3.540416 0.0007539585
#> 6  0     1 -3.533811 0.0007668695
```

``` r

describe_posterior(dfx, rvar_col = "my_rvar",
                   centrality = "MAP", ci_method = "hdi", ci = 0.8,
                   test = c("pd", "p_map", "rope"), rope_ci = 1, rope_range = c(-1, 0.5))
#> Summary of Posterior Distribution
#> 
#> mu   | sigma |  MAP |        80% CI | p (MAP) |     pd |          ROPE | % in ROPE
#> ----------------------------------------------------------------------------------
#> 0.00 |  1.00 | 0.12 | [-1.38, 1.15] |  0.997  | 50.30% | [-1.00, 0.50] |    53.10%
#> 0.50 |  0.50 | 0.52 | [-0.11, 1.17] |  0.634  | 84.78% | [-1.00, 0.50] |    49.75%
#> 1.00 |  0.25 | 1.00 | [ 0.70, 1.34] |  < .001 |   100% | [-1.00, 0.50] |     1.95%
```

``` r

bayesfactor_parameters(dfx, rvar_col = "my_rvar")
#> Warning: Prior not specified! Please specify priors (with column order matching
#>   'posterior') to get meaningful results.
#> Warning: Bayes factors might not be precise.
#>   For precise Bayes factors, sampling at least 40,000 posterior samples is
#>   recommended.
#> Bayes Factor (Savage-Dickey density ratio)
#> 
#> mu    | sigma |   BF
#> --------------------
#> 0.000 | 1.000 | 1.00
#> 0.500 | 0.500 | 1.00
#> 1.000 | 0.250 | 1.00
#> 
#> * Evidence Against The Null: 0
```

``` r

bayesfactor_parameters(dfx, rvar_col = "my_rvar", prior = "other_rvar")
#> Warning: Bayes factors might not be precise.
#>   For precise Bayes factors, sampling at least 40,000 posterior samples is
#>   recommended.
#> Bayes Factor (Savage-Dickey density ratio)
#> 
#> mu    | sigma |       BF
#> ------------------------
#> 0.000 | 1.000 |     1.01
#> 0.500 | 0.500 |    0.088
#> 1.000 | 0.250 | 1.34e-06
#> 
#> * Evidence Against The Null: 0
```

``` r

bayesfactor_restricted(dfx, rvar_col = "my_rvar",
                       hypothesis = "`x[1]` < `x[3]`")
#> Warning: Prior not specified! 
#>   Please specify priors (with column names matching 'posterior')
#>    to get meaningful results.
#> Bayes Factor (Order-Restriction)
#> 
#> Hypothesis      P(Prior) P(Posterior)   BF
#> `x[1]` < `x[3]`     0.83         0.83 1.00
#> 
#> * Bayes factors for the restricted model vs. the un-restricted model.
```

``` r

bayesfactor_restricted(dfx, rvar_col = "my_rvar", prior = "other_rvar",
                       hypothesis = "`x[1]` < `x[3]`")
#> Bayes Factor (Order-Restriction)
#> 
#> Hypothesis      P(Prior) P(Posterior)    BF
#> `x[1]` < `x[3]`     0.85         0.83 0.975
#> 
#> * Bayes factors for the restricted model vs. the un-restricted model.
```

``` r

si(dfx, rvar_col = "my_rvar", prior = "other_rvar", BF = c(1, 10))
#> Warning: Support intervals might not be precise.
#>   For precise support intervals, sampling at least 40,000 posterior
#>   samples is recommended.
#> Support Interval
#> 
#> Parameter |      BF = 1 SI |    BF = 10 SI
#> ------------------------------------------
#> x[1]      | [-3.94, -0.03] |              
#> x[2]      | [-1.44,  0.68] | [-1.44, 0.03]
#> x[3]      | [ 0.16,  1.28] | [ 0.16, 1.14]
```

<sup>Created on 2024-09-05 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

## 2DO

- [x] map_estimate
- [x] point_estimate
- [x] eti
- [x] hdi
- [x] spi
- [x] bci
- [x] ci
- [x] p_direction
- [x] p_map
- [x] p_rope
- [x] p_significance
- [x] rope
- [x] equivalence_test
- [x] describe_posterior
- [x] bayesfactor_parameters
- [x] bayesfactor_restricted
- [x] si
  - [x] ci printing methods (fixed in https://github.com/easystats/datawizard/commit/8ac3a93832f2ab7dc10b48bc6ff861a774e432c7 and https://github.com/easystats/datawizard/commit/e1086a4963c67ad4f539cf9599ff19c28ad2053e)
- [x] estimate_density
- [x] TESTS
- [x] DOCS
- [x] STYLER

### CODE REVIEW

File: R/bayesfactor_parameters.R [hunk 1]
```diff
@@ -185,7 +185,12 @@
 #' @author Mattan S. Ben-Shachar
 #'
 #' @export
-bayesfactor_parameters <- function(posterior, ...) {
+bayesfactor_parameters <- function(posterior,
+                                   prior = NULL,
+                                   direction = "two-sided",
+                                   null = 0,
+                                   ...,
+                                   verbose = TRUE) {
```
Comment (strengejacke) [inline R/bayesfactor_parameters.R:193]: I would use as few arguments as necessary for the generic. Else, _all_ other methods _at least_ have to include all these

Comment (mattansb) [reply R/bayesfactor_parameters.R:193]: Yes, these are the minimally required inputs and argument 😀

**Comment by mattansb** [comment]
I'm done - read to merge if you approve @strengejacke 👍

**Comment by strengejacke** [comment]
I fixed some lintrs. There seems to be a problem with the docs, see html-check that fails:

```
❯ checking HTML version of manual ... NOTE
  Found the following HTML validation problems:
  bayesfactor_parameters.html:300:38 (bayesfactor_parameters.Rd:245): Warning: nested emphasis <em>
  bayesfactor_restricted.html:200:38 (bayesfactor_restricted.Rd:138): Warning: nested emphasis <em>
  si.html:234:38 (si.Rd:166): Warning: nested emphasis <em>
```

can you look into this? Note that you need to pull changes first...

**Review by strengejacke — APPROVED** [review]

**Pull request merged**